File: src/main/java/de/pflugradts/pwman3/application/commandhandling/handler/RenameCommandHandler.java
package de.pflugradts.pwman3.application.commandhandling.handler;

import com.google.common.eventbus.Subscribe;
import com.google.inject.Inject;
import de.pflugradts.pwman3.application.UserInterfaceAdapterPort;
import de.pflugradts.pwman3.application.commandhandling.command.RenameCommand;
import de.pflugradts.pwman3.application.failurehandling.FailureCollector;
import de.pflugradts.pwman3.domain.model.transfer.Bytes;
import de.pflugradts.pwman3.domain.model.transfer.Input;
import de.pflugradts.pwman3.domain.model.transfer.Output;
import de.pflugradts.pwman3.domain.service.password.PasswordService;

import static de.pflugradts.pwman3.domain.service.password.PasswordService.EntryNotExistsAction.CREATE_ENTRY_NOT_EXISTS_EVENT;

public class RenameCommandHandler implements CommandHandler {

    @Inject
    private FailureCollector failureCollector;
    @Inject
    private UserInterfaceAdapterPort userInterfaceAdapterPort;
    @Inject
    private PasswordService passwordService;

    @Subscribe
    private void handleRenameCommand(final RenameCommand renameCommand) {
        if (passwordService.entryExists(renameCommand.getArgument(), CREATE_ENTRY_NOT_EXISTS_EVENT)
                .onFailure(throwable ->
                        failureCollector.collectPasswordEntryFailure(renameCommand.getArgument(), throwable))
                .getOrElse(false)) {
            final var secureInput = userInterfaceAdapterPort
                    .receiveSecurely(Output.of(Bytes.of("Enter new alias or nothing to abort: ")))
                    .onFailure(failureCollector::collectInputFailure)
                    .getOrElse(Input.empty());
            if (secureInput.isEmpty()) {
                userInterfaceAdapterPort.send(Output.of(Bytes.of("Empty input - Operation aborted.")));
            } else {
                passwordService.renamePasswordEntry(renameCommand.getArgument(), secureInput.getBytes())
                    .onFailure(failureCollector::collectRenamePasswordEntryFailure);
            }
            secureInput.invalidate();
        }
        userInterfaceAdapterPort.sendLineBreak();
        renameCommand.invalidateInput();
    }

}

File: src/main/kotlin/de/pflugradts/passbird/application/commandhandling/handler/egg/RenameCommandHandler.kt
package de.pflugradts.passbird.application.commandhandling.handler.egg
import de.pflugradts.passbird.application.UserInterfaceAdapterPort
import de.pflugradts.passbird.application.commandhandling.CommandExecutionTracker
import de.pflugradts.passbird.application.commandhandling.command.RenameCommand
import de.pflugradts.passbird.application.commandhandling.handler.TypedCommandHandler
import de.pflugradts.passbird.domain.model.egg.EggIdException
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.shellOf
import de.pflugradts.passbird.domain.model.transfer.Output.Companion.outputOf
import de.pflugradts.passbird.domain.model.transfer.OutputFormatting.OPERATION_ABORTED
import de.pflugradts.passbird.domain.service.password.PasswordService
import de.pflugradts.passbird.domain.service.password.PasswordService.EggNotExistsAction.CREATE_ENTRY_NOT_EXISTS_EVENT
class RenameCommandHandler constructor(
    private val passwordService: PasswordService,
    private val userInterfaceAdapterPort: UserInterfaceAdapterPort,
    private val commandExecutionTracker: CommandExecutionTracker,
) : TypedCommandHandler<RenameCommand>(RenameCommand::class.java) {
    override fun handleCommand(command: RenameCommand) {
        if (passwordService.eggExists(command.argument, CREATE_ENTRY_NOT_EXISTS_EVENT)) {
            val secureInput = userInterfaceAdapterPort.receive(outputOf(shellOf("Enter new EggId or nothing to abort: ")))
            if (secureInput.isEmpty) {
                commandExecutionTracker.markAborted()
                userInterfaceAdapterPort.send(outputOf(shellOf("Empty input - Operation aborted."), OPERATION_ABORTED))
            } else {
                try {
                    if (passwordService.renameEgg(command.argument, secureInput.shell).failure) {
                        commandExecutionTracker.markFailure()
                    }
                } catch (ex: EggIdException) {
                    commandExecutionTracker.markAborted()
                    userInterfaceAdapterPort.send(outputOf(shellOf("${ex.message} - Operation aborted."), OPERATION_ABORTED))
                }
            }
            secureInput.invalidate()
        } else {
            commandExecutionTracker.markFailure()
        }
        userInterfaceAdapterPort.sendLineBreak()
        command.invalidateInput()
    }
}
